infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/RestoreRedundancyMergeSegments.java

```diff
@@ -1,38 +1,213 @@
+                    if (RedundancyCalculator.canMergeSegments(newLayout)) {
+                        layoutManagementView.mergeSegments(newLayout);
+                    }
+                    // If the segments can't be merged, just propose a new layout.
+                    else {
+                        // Since we seal with a new epoch,
+                        // we also need to bump the epoch of the new layout.
+                        LayoutBuilder builder = new LayoutBuilder(newLayout);
+                        newLayout = builder.setEpoch(currentLayout.getEpoch() + 1).build();
+                        layoutManagementView
+                                .runLayoutReconfiguration(currentLayout, newLayout,
+                                        false);
+                    }
+                }
+                // Return the latest layout.
+                runtime.invalidateLayout();
+                return runtime.getLayoutView().getLayout();
+
+            } catch (WrongEpochException | QuorumUnreachableException | OutrankedException e) {
+                log.warn("Got: {}. Retrying: {} times.", e.getMessage(), retries.get());
+                if (retries.decrementAndGet() < 0) {
+                    throw new RetryExhaustedException("Retries exhausted.", e);
+                } else {
+                    throw new RetryNeededException();
+                }
+            } catch (TransferSegmentException e) {
+                throw new RetryExhaustedException("Transfer segment exception occurred.", e);
+            }
+
+        }).setOptions(retrySettings).run();
+
+    }
     /**
-     * Returns set of nodes which are present in the next index but not in the specified segment. These
-     * nodes have reduced redundancy and state needs to be transferred only to these before these segments can be
-     * merged.
+     * Sets the trim mark on this endpoint's log unit and also perform a prefix trim.
      *
-     * @param layout             Current layout.
-     * @param layoutSegmentIndex Segment to compare to get nodes with reduced redundancy.
-     * @return Set of nodes with reduced redundancy.
+     * @param layout  A current layout.
+     * @param runtime A current runtime.
+     * @return A retrieved trim mark.
      */
-    private Set<String> getNodesWithReducedRedundancy(Layout layout, int layoutSegmentIndex) {
-        // Get the set of servers present in the next segment but not in the this
-        // segment.
-        return Sets.difference(
-                layout.getSegments().get(layoutSegmentIndex + 1).getAllLogServers(),
-                layout.getSegments().get(layoutSegmentIndex).getAllLogServers());
+    long setTrimOnNewLogUnit(Layout layout, CorfuRuntime runtime) {
+
+        long trimMark = runtime.getAddressSpaceView().getTrimMark().getSequence();
+
+        Token prefixToken = new Token(layout.getEpoch(), trimMark - 1);
+        runtime.getLayoutView().getRuntimeLayout(layout)
```

Comment:
From like 203 to this point the epoch can change, but I think its "fine", since this rpc would fail with WrongEpochException, for the sake of consistency try to execute both requests on the same epoch (i.e. same layout)

Reply:
Fixed